common-api/src/main/java/org/pmiops/workbench/cdr/dao/DbDomainDao.java

```diff
@@ -0,0 +1,70 @@
+                "from db_domain d\n" +
+                "join concept c on d.domain_id = c.domain_id\n" +
+                "left join concept_synonym cs on c.concept_id=cs.concept_id \n" +
+                "where d.db_type = 'domain_filter' and (c.count_value > 0 or c.source_count_value > 0) \n" +
+                "and  ((((match(c.concept_name) against(?1 in boolean mode) ) or\n" +
+                "(match(cs.concept_synonym_name) against(?1 in boolean mode))) and\n" +
+                "c.standard_concept IN ('S', 'C') or (c.concept_id=?2 or c.concept_code=?2))\n" +
+                "group by d.domain_id, d.domain_display, d.domain_desc, d.db_type, d.concept_id\n" +
+                " \n" +
+                "union\n" +
+                " \n" +
+                "select d.domain_id, d.domain_display, d.domain_desc, d.db_type, d.domain_route,\n" +
+                "d.concept_id, count(q.concept_id) as standard_concept_count from db_domain d\n" +
+                "join achilles_results r on d.concept_id = r.stratum_1\n" +
+                "join concept q on r.stratum_2 = q.concept_id\n" +
+                "left join concept_synonym cs on q.concept_id=cs.concept_id \n" +
+                "where d.db_type = 'survey' and r.analysis_id = 3110\n" +
+                "and (((match(q.concept_name) against(?1 in boolean mode) ) or\n" +
+                "(match(cs.concept_synonym_name) against(?1 in boolean mode) ) or\n" +
+                "(q.concept_id=?2 or q.concept_code=?2)) or\n" +
+                "(match(r.stratum_4) against(?1 in boolean mode)))\n" +
+                "group by d.domain_id, d.domain_display, d.domain_desc, d.db_type, d.concept_id \n ")
+    List<DbDomain> findDomainSearchResults(String keyword,String query);
+
+    @Query(nativeQuery=true,value="select d.domain_id, d.domain_display, d.domain_desc, d.db_type,\n" +
+            "d.domain_route, d.concept_id, count(c.concept_id) as standard_concept_count from db_domain d\n" +
+            "join concept c on d.domain_id = c.domain_id\n" +
+            "where d.db_type = 'domain_filter' and c.count_value > 0 and c.standard_concept IN ('S', 'C')\n" +
+            "group by d.domain_id, d.domain_display, d.domain_desc, d.db_type, d.concept_id\n" +
+            " \n" +
+            "union\n" +
+            " \n" +
+            "select d.domain_id, d.domain_display, d.domain_desc, d.db_type, d.domain_route, d.concept_id,\n" +
+            "count(q.concept_id) as standard_concept_count from db_domain d\n" +
+            "join achilles_results r on d.concept_id = r.stratum_1\n" +
+            "join concept q on r.stratum_2 = q.concept_id\n" +
+            "where d.db_type = 'survey' and r.analysis_id = 3110\n" +
+            "group by d.domain_id, d.domain_display, d.domain_desc, d.db_type, d.concept_id order by db_type ASC")
+    List<DbDomain> findDomainTotals();
+
+    @Query(nativeQuery=true, value="select d.domain_id, d.domain_display, d.domain_desc, d.db_type,\n" +
```

Comment:
Shot in the dark, but perhaps try moving `saveConcepts()` / `saveDomains()` into the setUp method. I don't know what triggers transaction completion in these tests, but it's possible things are cached but not yet persisted when you're asking conceptDao about which ones exist.

Reply:
There was a typo in my SQL; I figured it out. :)

That said, I have discovered that this SQL doesn't perform very well. Using COUNT(DISTINCT) on c.concept_id interestingly runs faster in MySQL that without DISTINCT (not sure why), but takes around 7 seconds. Once you add in the SUM(c.standard_concept IN ('S', 'C')) it goes up to around 15 seconds -- so calculating counts for both standard and all concepts is pretty painful. 

We can cache these results in memory I guess, but there's going to be a big latency hit the first time we're calculating these counts. We could precalculate counts for the domains with no keyword filtering applied, which will help with the initial page load, but we can't do that once keyword filtering has been applied.

Maybe instead of getting both standard and all concept counts like I initially envisioned, we should just return one set of counts (for the current checkbox selection state) and have the client request standard vs. all concepts. 

Separately we should see if there's anything we can do to make the SQL go faster. I'm going to see if a compound index helps.

Reply:
Adding a compound index on (domain_id, standard_concept) brings this down to around 8 seconds (whether or not standard counts are generated). So I'm going to add that to this PR.

We should still talk about caching strategies so we don't take this hit so often.

Reply:
That's an awesome improvement.  +1 on getting the selection state only.

Reply:
I had forgotten the filter on c.count_value > 0; once I added that and made the index include count_value, things are snappier.